File: java/_MinimumWindowSubstring.java
import java.util.HashMap;

/**

 Given a string S and a string T, find the minimum window in
 S which will contain all the characters in T in complexity O(n).

 For example,
 S = "ADOBECODEBANC"
 T = "ABC"
 Minimum window is "BANC".

 Note:
 If there is no such window in S that covers all characters in T,
 return the empty string "".
 If there are multiple such windows, you are guaranteed that there
 will always be only one unique minimum window in S.

 */

public class _MinimumWindowSubstring {
    public String minWindow(String source , String target) {
        // write your code here

        int left = 0;
        int right = 0;
        int min = Integer.MAX_VALUE;
        int[] map = new int[256];
        int startIndex = 0;
        int total = 0;

        for (int i = 0; i < target.length(); i++) {
            map[target.charAt(i)]++;
            total++;
        }

        for (left = 0; left < source.length(); left++) {
            while (right < source.length() && total > 0) {
                if (map[source.charAt(right)] > 0) {
                    total--;
                }
                map[source.charAt(right)]--;
                right++;
            }
            if (total == 0 && (right - left) < min) {
                startIndex = left;
                min = right - left;
            }

            if (map[source.charAt(left)] >= 0) {
                total++;
            }
            map[source.charAt(left)]++;
        }

        return min == Integer.MAX_VALUE ? "" :
                source.substring(startIndex, startIndex + min);
    }


    public String minWindow3(String s, String t) {
        int[] cnt = new int[128];

        // count all t's appearence.
        // when total is zero, it means it is valid substring
        for (char c : t.toCharArray()) {
            cnt[c]++;
        }

        int from = 0;
        int total = t.length();
        int min = Integer.MAX_VALUE;

        for (int right = 0, left = 0; right < s.length(); right++) {
            // ====== handle right pointer
            if (cnt[s.charAt(right)] > 0) {
                // when we find a target character, we decrease total.
                total--;
            }
            cnt[s.charAt(right)]--;


            // ====== start to deal with left pointer
            // If total is still zero, we keep move left pointer.
            while (total == 0) {
                // find a new minimum
                if (right - left  +  1 < min) {
                    min = right - left + 1;
                    from = left;
                }

                cnt[s.charAt(left)]++;
                if (cnt[s.charAt(left)] > 0) {
                    total++;
                }
                left++;
            }

        }

        return (min == Integer.MAX_VALUE) ? "" : s.substring(from, from + min);
    }

    public String minWindow2(String s, String target) {
        HashMap<Character, Integer> targetCharMap = new HashMap();

        // ceate hashmap with all possible value;
        for (char c : s.toCharArray()) {
            targetCharMap.put(c, 0);
        }

        // Track all the appearance in target arrary
        for (char c : target.toCharArray()) {
            if (targetCharMap.containsKey(c)) {
                targetCharMap.put(c, targetCharMap.get(c) + 1);
            } else {
                return "";
            }
        }

        int start = 0, end = 0, targetCount = target.length();
        int resultStartIndex = 0;
        int result = Integer.MAX_VALUE;

        while(end < s.length()) {
            char c1 = s.charAt(end);
            if (targetCharMap.get(c1) > 0) {
                targetCount--;
            }

            targetCharMap.put(c1, targetCharMap.get(c1) - 1);
            end++;

            // Move left index to find shorter result.
            while (targetCount == 0) {
                if (result > end - start) {
                    result = end - start;
                    resultStartIndex = start;
                }

                char c2 = s.charAt(start);
                targetCharMap.put(c2, targetCharMap.get(c2) + 1);

                if (targetCharMap.get(c2) > 0) {
                    targetCount++;
                }

                start++;
            }
        }

        return result == Integer.MAX_VALUE ? "" :
                s.substring(resultStartIndex, resultStartIndex + result);
    }
}

File: algorithms/sliding-window/minimum-window-substring/minimum-window-substring.ts
/**
https://leetcode.com/problems/minimum-window-substring/submissions/

76. Minimum Window Substring
Given two strings s and t of lengths m and n respectively, return the minimum window substring of s such that every character in t (including duplicates) is included in the window. If there is no such substring, return the empty string "".

The testcases will be generated such that the answer is unique.

A substring is a contiguous sequence of characters within the string.



Key:
Track character appearance count;
Track valid substring (using count;)
 */
function minWindow(s: string, t: string): string {
  const map = new Map();
  let result;
  // Count equal to zero represents it is a valid string.
  let count = 0;

  for (let i = 0; i < t.length; i++) {
    if (map.has(t[i]) === false) {
      map.set(t[i], 0);
    }

    map.set(t[i], map.get(t[i]) + 1);
    count += 1;
  }

  let begin = 0;
  let end = 0;
  let head = 0;
  let distance = Number.MAX_SAFE_INTEGER;

  while (end < s.length) {
    if (map.has(s[end])) {
      // Only decrease count when there is a character needs to be removed.
      if (map.get(s[end]) > 0) {
        count -= 1;
      }

      map.set(s[end], map.get(s[end]) - 1);
    }

    while (count === 0) {
      if (end - begin + 1 < distance) {
        distance = end - begin + 1;
        head = begin;
      }

      if (map.has(s[begin])) {
        // Only increase count when we need the characters.
        if (map.get(s[begin]) === 0) {
          count += 1;
        }
        map.set(s[begin], map.get(s[begin]) + 1);
      }
      begin += 1;
    }

    end += 1;
  }

  // console.log(begin);
  // console.log(distance);
  // console.log(result === undefined)
  return distance === Number.MAX_SAFE_INTEGER
    ? ""
    : s.substring(head, head + distance);
}

// console.log(minWindow("ADOBECODEBANC", "ABC")); // "BANC"
// console.log(minWindow("a", "a")); // "a"
// console.log(minWindow("a", "aa")); // "a"
